cmd/ingester/app/consumer/deadlock_detector.go

```diff
@@ -0,0 +1,127 @@
+	return deadlockDetectorFactory{
+		metricsFactory: factory,
+		logger:         logger,
+		interval:       interval,
+		panicFunc: func(partition int32) {
+			factory.Counter("deadlockdetector.panic-issued", map[string]string{"partition": strconv.Itoa(int(partition))}).Inc(1)
+			time.Sleep(time.Second) // Allow time to flush metric
+
+			buf := make([]byte, 1<<20)
+			logger.Panic("No messages processed in the last check interval",
+				zap.Int32("partition", partition),
+				zap.String("stack", string(buf[:runtime.Stack(buf, true)])))
+		},
+	}
+}
+
+// startMonitoringForPartition monitors the messages consumed by the partition and signals for the partition to by
+// closed by sending a message on the closePartition channel.
+//
+// Closing the partition should result in a rebalance, which alleviates the condition. This means that rebalances can
+// happen frequently if there is no traffic on the Kafka topic. This shouldn't affect normal operations.
+//
+// If the message send isn't processed within the next check interval, a panic is issued.This hack relies on a
+// container management system (k8s, aurora, marathon, etc) to reschedule
+// the dead instance.
+//
+// This hack protects jaeger-ingester from issues described in  https://github.com/jaegertracing/jaeger/issues/1052
+//
+func (s *deadlockDetectorFactory) startMonitoringForPartition(partition int32) *deadlockDetector {
+	var msgConsumed uint64
+	w := &deadlockDetector{
+		msgConsumed:    &msgConsumed,
+		ticker:         time.NewTicker(s.interval),
+		closePartition: make(chan struct{}, 1),
+		done:           make(chan struct{}),
+		logger:         s.logger,
+	}
+
+	go func() {
+		for range w.ticker.C {
+			select {
+			case <-w.done:
+				s.logger.Info("Closing ticker routine", zap.Int32("partition", partition))
+				return
+			default:
+				if atomic.LoadUint64(w.msgConsumed) == 0 {
+					select {
+					case w.closePartition <- struct{}{}:
+						s.logger.Warn("Signalling partition close due to inactivity", zap.Int32("partition", partition))
+					default:
+						// If closePartition is blocked, the consumer might have deadlocked - kill the process
+						s.panicFunc(partition)
+					}
+				} else {
+					atomic.StoreUint64(w.msgConsumed, 0)
+				}
+			}
+		}
+	}()
+
+	return w
+}
+
+// startMonitoring is to monitor that the sum of messages consumed across all partitions is non zero for the given interval
+// If it is zero when there are producers producing messages on the topic, it means that sarama-cluster hasn't
+// retrieved partition assignments. (This case will not be caught by startMonitoringForPartition because no partitions
+// were retrieved).
+func (s *deadlockDetectorFactory) startMonitoring() *deadlockDetector {
+	return s.startMonitoringForPartition(-1)
+}
+
+func (w *deadlockDetector) getClosePartition() chan struct{} {
+	return w.closePartition
+}
+
+func (w *deadlockDetector) close() {
+	w.logger.Info("Closing deadlock detector")
```

Comment:
nit: better to capture partition # in the detector and include it in the log stmt